runtime/participant-runtime.ss

```diff
@@ -767,3 +767,380 @@
+                                       ; ensure this is handled at cli options level.
+
+(def (send-contract-handshake/stdout self handshake)
+  (def io-context (.@ self io-context))
+  (.call io-context send-handshake handshake))
+
+(def (send-contract-handshake/libp2p libp2p-client dest-address handshake)
+  (displayln MAGENTA "Sending handshake to multiaddr..." END)
+  (def handshake-string (string<-json (json<- AgreementHandshake handshake)))
+  (dial-and-send-contents libp2p-client dest-address handshake-string)
+  (displayln)
+  (force-output))
+
+
+;; ------------------ Listen for handshake
+
+
+;; : AgreementHandshake <- Runtime
+(def (listen-for-handshake self)
+  (def channel (.@ self off-chain-channel))
+  (match (.@ channel tag)
+    ('stdio
+     (let ()
+       (displayln MAGENTA "Listening for handshake via stdin ...")
+       (listen-for-handshake/stdio self)))
+    ('libp2p
+     (let ()
+       (displayln MAGENTA "Listening for handshake via libp2p ...")
+       (def handshake-str (.call channel poll-buffer))
+       (displayln MAGENTA "Received handshake")
+       (def handshake (<-json AgreementHandshake (json<-string handshake-str)))
+       handshake))
+    (else (error "Invalid channel"))))
+
+(def (listen-for-handshake/stdio self)
+  (def io-context (.@ self io-context))
+  (.call io-context receive-handshake))
+
+
+(def (listen-for-handshake/libp2p libp2p-client)
+  (let* ((self (libp2p-identify libp2p-client)))
+    (for (p (peer-info->string* self))
+      (displayln "I am " p))
+    (displayln "Listening for incoming connections")
+    ;; TODO: use parameterize instead?
+    (def ret (make-parameter #f))
+    (libp2p-listen libp2p-client [chat-proto] (chat-handler ret))
+    ;; TODO close the connection
+    (let loop ()
+      (or (ret)
+          (begin
+            (thread-sleep! 3)
+            (loop))))))
+
+
+;; ------------------ Libp2p client methods
+
+
+(def (chat-writer s contents)
+  (display "> ")
+  (bio-write-string contents (stream-out s))
+  (bio-write-char #\newline (stream-out s))
+  (bio-force-output (stream-out s)))
+
+;; peer-multiaddr-str: destination multiaddress,
+;; NOTE: has the constraint that it needs to contain a peerId,
+;; so we can verify the recipient's identity.
+;;
+;; host-addresses: Multi addresses this participant listens to on their host machine.
+;; contents: string
+(def (dial-and-send-contents libp2p-client dest-address-str contents)
+  (let* ((self (libp2p-identify libp2p-client))
+         (peer-multiaddr (string->peer-info dest-address-str)))
+    (for (p (peer-info->string* self))
+      (displayln "I am " p))
+    (displayln "Connecting to " dest-address-str)
+    (libp2p-connect/poll libp2p-client peer-multiaddr timeout: 60)
+    (let (s (libp2p-stream libp2p-client peer-multiaddr [chat-proto]))
+      (chat-writer s contents))))
+
+;; polls every 1s if other party is offline / unreachable, until timeout
```

Comment:
Add comment:
If we try to poll when running the interaction via the cli, it works just fine.
In integration tests however, the following error message comes up:
```
 *** ERROR IN vyzo/libp2p/client#libp2p-stream__% -- libp2p-request: [libp2p-error] protocol not supported
```

Will require more investigation to understand under what scenarios this comes up.

Reply:
Have added a fixme to the test, detailing exactly what happens.